Unified reload script

Dridi opened the pull request:
⚠️ WORK IN PROGRESS ⚠️ 

Following offline discussions and observations made in #30, this is a new script that should work on all supported platforms. This isn't finished but if you have early comments, don't hesitate.

Goals:

- narrow scope: reload VCL from service manager
- self-contained: no coupling to /etc, only the `varnish-cli`
- limited interface
- simple defaults

Non goals:

- support platforms outside pkg-varnish-cache [1]
- support older Varnish versions [2]
- support existing scripts features
- handle all `varnishd` and `varnishadm` options

This will hopefully remove future problems akin to #35, #51, #61 and #64. The first prototype has seen limited testing.

[1] but I try to keep the script portable
[2] but if it can work, let's help it

denisbr commented:
I have reviewed this with @Dridi and I'm in favour of this approach, and keeping the list of goals focused.

Dridi commented:
I believe the script is now feature-complete for the goal I initially set. It was successfully tested with my insane^Wlittle collection of [mostly-]posix-compliant shells (although only `bash` and `dash` matter here) with Varnish 5.0 because that's what shipped on my system.

I will extend testing to more corner cases and Varnish versions (ideally >= 4.1). In particular, the remaining XXX comment.

Dridi commented:
Forgot another feature I had strongly considered for this script: keeping a maximum number of available reloads.

Dridi commented:
I wrote a test case ([test.vtc.txt](https://github.com/varnishcache/pkg-varnish-cache/files/860773/test.vtc.txt)) to confirm `vcl.show`'s behavior when multiple VCLs share the same name. Only the available/active (or the other way around not-discarded) VCL is shown. This way I could get rid of useless code in 8ffcb1c. Tested with all releases since 4.1.3.

And on the same set of releases I also checked the output of `vcl.show` and the expectations of the `case` statement: all green. I'd expect the reload script to also work with all the releases since 4.1.3 (possibly older too, but I was limited by the test case) but I plan to make it a tad more reliable.

Dridi commented:
Quick check again, the `case` statement wouldn't work with EOL'd 4.0 releases.

Dridi commented:
I think this time it's feature complete, ready for discussion on how to integrate it in service managers. And of course ready for extensive testing.

Dridi commented:
@RyanJarv, I requested reviews from downstream packagers @ingvarha and @ssm but feel free to try it in your chef integration and let us know if you find any problem with it.

ssm commented:
I like it.  It works for me, and the script is readable.

I tried running with '-h' to display help, but that was sent on as the filename to read from.

```
# ./varnishreload -h
Command: varnishadm -n '' -- vcl.load reload_20170414_195831 -h

Command failed with error code 106
Message from VCC-compiler:
Cannot read file '-h' (No such file or directory)
Running VCC-compiler failed, exited with 2
VCL compilation failed
```

I took a stab at addressing that, at https://gist.github.com/ssm/7af538c61e9905fa4bf0b59818ff25df

My proposed changes are:

* Use "getopts" to parse options. This will emit usage and exit in case of unrecognized options. You can  use '--' to indicate end of options, in case the file name starts with '-'.
* Use "case" to handle file argument
* Expand file argument with "readlink" in case the path to the file given is relative. Also check if the file is present.
* Add "-h" option to emit help

Dridi commented:
Hello @ssm, thanks a lot for the review and the suggestion. Seeing how your first try was to run it with `-h` and how sometimes I do the same I added support for the help option and  moved arguments parsing to getopts.

> Use "case" to handle file argument

I didn't understand this comment so I had to look at your script to figure what you meant. And I'm more satisfied with a single concise `test` than a more involved `case`. See also below for other problems with that `case`.

> Expand file argument with "readlink" in case the path to the file given is relative.

I don't want to prevent users from reloading relative files, simply because Varnish supports it and will search for them in the `vcl_path`.

> Also check if the file is present.

Since `vcl.load` will fail if the file is missing, I don't need to encumber the script with more error handling. Especially if the file name is relative, I do not wish to implement a path search in shell (although I sense much fun in that direction).

> I like it. It works for me, and the script is readable.

Thanks! Please have a look at the latest revision with changes made after your comment.

ssm commented:
Dridi Boukelmoune <notifications@github.com> writes:

> Hello @ssm, thanks a lot for the review and the suggestion. Seeing how
> your first try was to run it with `-h` and how sometimes I do the same
> I added support for the help option and moved arguments parsing to
> getopts.

[...]

> I don't want to prevent users from reloading relative files, simply
> because Varnish supports it and will search for them in the
> `vcl_path`.

That makes sense, I did not consider the vcl_path at all, which makes
some of my proposed changes not work well.

A few notes about the help output:

For "<file>", it may be helpful to say how varnishreload parses that
argument, if present. Something like:

  The optional <file> argument refers to a VCL file. You may specify an
  absolute path to this file, or a relative path to be looked up in the
  VCL Path. ("varnishadm param.show vcl_path")

For "-n", the usage output says "directory", but it matches the "-n"
parameter of varnishadm and varnishd, which describes it to be "instance
name", (which could be an absolute path to a directory).  Would it make
sense to match the description from the man pages of those commands?

-- 
Stig Sandbeck Mathisen
https://fnord.no/

RyanJarv commented:
Noticed is the debian init script uses the quiet option in the init.d script. Made a PR to add that to this script. https://github.com/Dridi/pkg-varnish-cache/pull/1

One other thing, we include the reload script in the apache2 licensed chef cookbook to work around issues with the distro's version. I think the current redhat version of this script includes a reference to the BSD license for a that reason, can we get that added here as well?

Other then that it looks good to me.

RyanJarv commented:
Oh I did notice this breaks on varnish 4.0, not sure if that is a concern or not since it's EOL currently. 

vcl.show on the boot config show's 'input' in the place of the file path on other versions.

Dridi commented:
@ssm, if we are down to discussing wording in the usage message, I'm glad! I'm happy to take any suggestion in this area, I didn't put much effort as I was looking for a working script first. I will push an update with your remarks.

@RyanJarv said:
> [...] BSD license [...] can we get that added here as well?

Yes I will add the license to the script.

> Noticed is the debian init script uses the quiet option in the init.d script.

The goal (once we agree on the script) is to then adapt our services (sysvinit and systemd) to use that script on both debian- and rhel-like systems. It's also expressed in the non-goals that we aren't taking features from existing scripts but settle on the bare minimum for smooth integration in service managers. A redirect to `/dev/null` sounds straightforward enough to me :)

> Oh I did notice this breaks on varnish 4.0, not sure if that is a concern or not since it's EOL currently.

Yes, would have been nice to have but I'm not putting effort in supporting an EOL series.

denisbr commented on varnishreload:
```diff
@@ -10,14 +10,17 @@ WARMUP=
+	-h           : show this help and exit
 	-m <maximum> : maximum number of available reloads to leave behind
 	-n <workdir> : for a different working directory for varnishd
 	-w <warmup>  : the waiting period between load and use operations
```
"the waiting period (seconds) between load and use operations" maybe?

Dridi commented:
I think all feedback is now taken care of, please have another look. I would love to hear from @ingvarha too, even though we work on the same platforms.

Dridi commented:
To spice things up a bit I tried it on FreeBSD (default shell is `csh`, not using GNU versions of `sed` and `awk`...) and it only partially works.

Dridi commented:
`awk` on FreeBSD doesn't seem to support the `{num}` syntax:

```
$NF ~ /^reload_[0-9]{8}_[0-9]{6}$/'
```

This is less accurate but works:

```
$NF ~ /^reload_[0-9]+_[0-9]+$/
```

I will see whether I can get something better, otherwise that'll do.

edit: according to the `awk` manual on freebsd, regular expressions are of the extended persuasion, same as `egrep`. According to the `grep` manual extended regex supports the `{n}` notation, and `egrep ^reload_[0-9]{8}_[0-9]{6}$` works as expected. I will not try to figure what I'm doing wrong, the less accurate regex will do just fine.

Dridi commented:
I found a bug in the script, fixed in 6f6dfea. And with f003c72 it works on FreeBSD too. So apparently only `awk` usage didn't fully work there, despite the docs hinting that it should.

I'm getting better at portable shell scripting :)

rezan commented:
Running into the following problem. Im testing with the latest script as of the time of this post and `varnish-4.1.5`. I have started varnishd using a relative path to my VCL. When I run the reload script without any parameters:

```
> ./varnishreload
Error: active VCL file not found (got test.vcl)
```

When I run passing in `test.vcl`, it works:

```
> ./varnishreload test.vcl
VCL compiled.
VCL 'reload_20170607_211921' now active
```

Here is the `vcl.show -v [active]` comment for the original boot vcl which I tried reloading:

```
// VCL.SHOW 0 321 test.vcl
```

I am guessing this check is meant for something, but its failing for valid VCL files with relative paths.

Dridi commented:
I don't get it, I consistently get an absolute path even when I load a relative file:

```
$ /opt/varnish-4.1.5/bin/varnishadm -n /tmp vcl.load test default.vcl
VCL compiled.

$ /opt/varnish-4.1.5/bin/varnishadm -n /tmp vcl.use test
VCL 'test' now active
$ /opt/varnish-4.1.5/bin/varnishadm -n /tmp vcl.show -v test | sed 1q
// VCL.SHOW 0 69 /opt/varnish-4.1.5/etc/varnish/default.vcl
$ ./varnishreload -n /tmp
VCL compiled.
VCL 'reload_20170608_073615' now active
```

That's why the script is expecting absolute paths only.

rezan commented:
Hmm... Can you try adding `-d` when you start varnishd? That's how I was running this.

Dridi commented:
I was already starting varnish in debug (`-d`) mode, but after discussing this on IRC I was finally able to reproduce it, and you actually uncovered a bug in `varnishd`, not this script.

The `-f` option in Varnish 4.1 doesn't behave like `vcl.load`, and disregards the `vcl_path` (or `vcl_dir`) and if I run `varnishd` from the `/etc/varnish` directory with a default install, I'm able to use the relative path `default.vcl`.

What's more interesting is that since `-f` does not honor the `vcl_path`, the file name is never resolved to an absolute path, so this is a side effect of a bug in `varnishd`:

```
$ file /etc/varnish/default.vcl 
/etc/varnish/default.vcl: ASCII text
$ varnishd -d -a :0 -p vcl_path=/etc/varnish -f default.vcl -n /tmp
Error: Cannot read -f file (default.vcl): No such file or directory
```

I will also need to see how 5.1 handles it since the revamp of the `-f` option (allowed multiple times, with a special meaning for an empty file name...) but this difference in behavior is not documented so I think it simply was overlooked.

gauthier-delacroix commented:
@Dridi I've (finally) tried your reload script with 4.1.8 and it works fine.

Compared to the one I've submitted in #9 (and use in production since then), yours only disards `reload_...`  configs: This makes sense, other ones being boot and manually loaded configs that shouldn't be taken in the rotation.

I have 2 questions:
- Do you plan to call this script from service files (as you didn't change them in the PR) or would we have to call it directly ?
- If merged, is there a chance to see it packaged with a future 4.1.x release ? (not sure how this repo is  linked to Varnish releases)

Both aren't issues as I'm already customising service files through Varnish, but less custom is always better (as discussed in Dridi/libvmod-querystring#30 :wink:)

Thanks !

Dridi commented:
Regarding my last comment, the `-f` bug was fixed in varnishcache/varnish-cache#2342.

@gauthier-delacroix the plan is to ideally get this script in upstream Varnish, so that we don't end up with the 5.0 problems with scripts from this repo no longer working. If it lives in Varnish, it would be checked for regressions automatically via the test suite (someone would have to write the test cases=).

Whether it lands in Varnish or stays confined in the packaging, the plan is to supersede the existing scripts, retire them, and update service definitions accordingly. As it stands, the script is perfectly fine, except that it doesn't handle the case where the active VCL would be a label and I will likely not do that before the 6.0 release.

gauthier-delacroix commented:
I'm still on 4.1 so I didn't try VCL labels yet (I use a lot of includes to handle virtual hosts so it's a feature I'd really like to play with) but if I understand correctly, since labels are only meant to be called by active VCL, the active VCL has no reason to be labeled...

Maybe varnishd should enforce that active VCL **can't** be labeled, and that a label can't be set active.

Dridi commented:
Switching to labels is one side of the feature, the first one is "symbolic linking". You may say that the active VCL is the `production` label, load new VCL and label it `production`. This will in effect reload the active VCL with an indirection. You may have another `maintenance` label that fails health checks in order to be "removed" from your infrastructure (load balancer, name server etc).

So when doing simple operations, you can move `production` and `maintenance` by simply using the labels and not care about the actual name (let's say `reload_DDDDDDDD_TTTTTT`). This was not in the initial description/plan for this pull request.

New goal:

* proper reload when active VCL is a label

New non goal:

* reloading labels that the active VCL may switch to

gauthier-delacroix commented:
@Dridi I tried to replace the old `/usr/sbin/varnish_reload_vcl` by your script in the service file (rhel/systemd) and I notice an issue with using instance name to access admin: The old script was using admin listen address/port which are available in the service environment. The instance name is not.
That means that to use your script from a service file, you either have to variabilize the instance name, which is difficult since instances aren't named by default (I do name mine), or you have to....customize the service file if you're using named instance.

Is there any specific reason why you're using instance name instead of address/port ? Or do you think it could be changed ?

Dridi commented:
This is by design. All you need is enough privileges (which should be guaranteed by your service manager) and the `-n` option (optional). `varnishadm` will then find all it needs automatically. We can discuss service integration directly via email or IRC if you want and I will post the conclusion here. For IRC you should find me on Varnish's server as dridi during business hours (I believe we're in the same timezone).

Dridi commented:
It's been a long time but yesterday I spent two hours of quality time in a train and figured that since I was on a shell scripting marathon it was a good time to revisit this.

The good news:
- it works with labels, I will submit it in Varnish Cache itself
- a lot less `varnishadm` executions are made
- overall, less purity but also less complexity
- it seems to be compatible with 4.1 too (I just tested that a bit)

The bad news:
- it needs testing again from interested parties
- I'll need to write a test suite for inclusion in Varnish Cache (more delays)

Currently it's been tested on Fedora, so that leaves us all platforms we care about to test it on:
- [ ] FreeBSD (me)
- [ ] SunOS (me)
- [ ] Ubuntu 14.04
- [ ] Ubuntu 16.04
- [ ] Debian 8
- [ ] Debian 9
- [ ] el6
- [ ] el7

The best is to test it with Varnish 5.1 or later to exercise all features, for Varnish 4.1 we will have the test suite once available. The important part is the stability of the shell script itself. Please give it a try, and let us know if it works for you and on which platform, even if it isn't listed above.

Napsty commented:
Just tested https://raw.githubusercontent.com/Dridi/pkg-varnish-cache/c8ab26809eea88cb73eed1bdee7ba55356e6fdf4/varnishreload with Varnish 5.2.0-1 on Ubuntu Xenial.

	root@onl-lb01-t:~# systemctl reload varnish-titan
	Job for varnish-titan.service failed because the control process exited with error code. See "systemctl status varnish-titan.service" and "journalctl -xe" for details.

	root@onl-lb01-t:~# systemctl status varnish-titan.service
	● varnish-titan.service - Varnish HTTP accelerator
	   Loaded: loaded (/etc/systemd/system/varnish-titan.service; enabled; vendor preset: enabled)
	   Active: active (running) (Result: exit-code) since Fri 2017-09-22 15:42:19 CEST; 2 weeks 5 days ago
	     Docs: https://www.varnish-cache.org/docs/4.1/
		   man:varnishd
	  Process: 23802 ExecReload=/usr/sbin/varnishreload (code=exited, status=1/FAILURE)
	 Main PID: 866 (varnishd)
	    Tasks: 217
	   Memory: 165.2M
	      CPU: 19min 51.047s
	   CGroup: /system.slice/varnish-titan.service
		   ├─ 866 /usr/sbin/varnishd -j unix,user=vcache -n varnish-titan -F -a :6081 -T localhost:6082 -f /etc/varnish/titan.vcl -S /etc/varnish/secret -s malloc,4096m
		   └─1328 /usr/sbin/varnishd -j unix,user=vcache -n varnish-titan -F -a :6081 -T localhost:6082 -f /etc/varnish/titan.vcl -S /etc/varnish/secret -s malloc,4096m

	Oct 12 11:08:36 onl-lb01-t systemd[1]: Reloading Varnish HTTP accelerator.
	Oct 12 11:08:41 onl-lb01-t varnishreload[23802]: Command: varnishadm -n '' -- vcl.list
	Oct 12 11:08:41 onl-lb01-t varnishreload[23802]: .....
	Oct 12 11:08:41 onl-lb01-t varnishreload[23802]: Could not get hold of varnishd, is it running?
	Oct 12 11:08:41 onl-lb01-t varnishreload[23802]: Error: failed to get the VCL list
	Oct 12 11:08:41 onl-lb01-t systemd[1]: varnish-titan.service: Control process exited, code=exited status=1
	Oct 12 11:08:41 onl-lb01-t systemd[1]: Reload failed for Varnish HTTP accelerator.
	Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.

I'm unsure about the correct usage of the reload script. Do you have an example for the ExecReload line in the systemd unit file?
I also tried "ExecReload=/usr/sbin/varnishreload -l varnish-titan" with the same result.

Dridi commented:
@Napsty you need to match the `-n varnish-titan` argument with `varnishd`'.

Napsty commented:
Yes, with -n varnish-titan in ExecReload it seems to work. I was unsure about that because in the help output it mentions "workdir". Maybe make a minor text change to avoid misunderstandings, e.g.:

    -n <workdir> : specify the instance name or directory for the varnishd instance

Reload seems to work when correctly used ;)

	root@onl-lb01-t:~# systemctl reload varnish-titan

	root@onl-lb01-t:~# systemctl status varnish-titan.service
	● varnish-titan.service - Varnish HTTP accelerator
	   Loaded: loaded (/etc/systemd/system/varnish-titan.service; enabled; vendor preset: enabled)
	   Active: active (running) (Result: exit-code) since Fri 2017-09-22 15:42:19 CEST; 2 weeks 5 days ago
	     Docs: https://www.varnish-cache.org/docs/4.1/
		   man:varnishd
	  Process: 27654 ExecReload=/usr/sbin/varnishreload -n varnish-titan (code=exited, status=0/SUCCESS)
	 Main PID: 866 (varnishd)
	    Tasks: 217
	   Memory: 165.4M
	      CPU: 19min 51.701s
	   CGroup: /system.slice/varnish-titan.service
		   ├─ 866 /usr/sbin/varnishd -j unix,user=vcache -n varnish-titan -F -a :6081 -T localhost:6082 -f /etc/varnish/titan.vcl -S /etc/varnish/secret -s malloc,4096m
		   └─1328 /usr/sbin/varnishd -j unix,user=vcache -n varnish-titan -F -a :6081 -T localhost:6082 -f /etc/varnish/titan.vcl -S /etc/varnish/secret -s malloc,4096m

	Oct 12 11:18:11 onl-lb01-t varnishd[866]: CLI telnet ::1 54142 ::1 6082 Wr 200 VCL compiled.
	Oct 12 11:18:11 onl-lb01-t varnishreload[27654]: VCL 'reload_20171012_111810' compiled
	Oct 12 11:18:11 onl-lb01-t varnishd[866]: CLI telnet ::1 54150 ::1 6082 Rd auth a9b970cd1728ad340402bdaeaa4226ca9fb31114b1a6e4f8edf667ca8f780f93
	Oct 12 11:18:11 onl-lb01-t varnishd[866]: CLI telnet ::1 54150 ::1 6082 Wr 200 -----------------------------
		                                  Varnish Cache CLI 1.0
		                                  -----------------------------
		                                  Linux,4.4.0-92-generic,x86_64,-junix,-smalloc,-smalloc,-hcritbit
		                                  varnish-5.2.0 revision 4c4875cbf
		                                  
		                                  Type 'help' for command list.
		                                  Type 'quit' to close CLI session.
	Oct 12 11:18:11 onl-lb01-t varnishd[866]: CLI telnet ::1 54150 ::1 6082 Rd ping
	Oct 12 11:18:11 onl-lb01-t varnishd[866]: CLI telnet ::1 54150 ::1 6082 Wr 200 PONG 1507799891 1.0
	Oct 12 11:18:11 onl-lb01-t varnishd[866]: CLI telnet ::1 54150 ::1 6082 Rd vcl.use reload_20171012_111810
	Oct 12 11:18:11 onl-lb01-t varnishd[866]: CLI telnet ::1 54150 ::1 6082 Wr 200 VCL 'reload_20171012_111810' now active
	Oct 12 11:18:11 onl-lb01-t varnishreload[27654]: VCL 'reload_20171012_111810' now active
	Oct 12 11:18:11 onl-lb01-t systemd[1]: Reloaded Varnish HTTP accelerator.

Also tested with 4.1.7-1 on Ubuntu Xenial.

Dridi commented:
Yeah, it's a bit mystifying but `-n` is actually a directory, the argument can either be a relative or an absolute path but ultimately it's the working directory. The rule of thumb when using a `varnish*` tool is to always match the `-n` option with `varnishd`'s. Thanks for testing on Xenial!

Napsty commented:
Successfully tested with Varnish 5.0.0-7 on Debian 9 (stretch). 
Successfully tested with Varnish 5.2.0-1 on Debian 9 (stretch).

Napsty commented:
@Dridi once successfully tested and approved, will the new varnishreload script be part of the "varnish" package? What about the varnish packages in the default repos of Ubuntu and Debian? Will the maintainer update the packages there, too?

Dridi commented:
> @Dridi once successfully tested and approved, will the new varnishreload script be part of the "varnish" package?

Yes, when I suggested that, the condition was to have a proper support for labels. Now it's suitable for a submission.

> What about the varnish packages in the default repos of Ubuntu and Debian? Will the maintainer update the packages there, too?

Until a release includes `varnishreload` it's up to them, depending on their package maintenance policy. Considering that some already ship a reload script from this repo, the only roadblock would be the risk of breaking user installations that rely on the old script.

Is the new script running in production? If not, do you plan to use it?

Napsty commented:
> Is the new script running in production? If not, do you plan to use it?

Not yet in production, but will be deployed to in the next days.

ssm commented:
As soon as it is merged, I'll add it to the Debian packaging.

Dridi commented:
@ssm good to know! And while I have your attention, can you test the new version of the script on Jessie? And if I'm not stretching too hard, Ubuntu 14.04 too?

ssm commented on varnishreload:
```diff
@@ -0,0 +1,216 @@
+}
+
+vcl_reload_name() {
+	printf "reload_%s" "$(date +%Y%m%d_%H%M%S)"
```
You might want a _%N at the end of the format string, otherwise you can't reload twice during the same second.  (My autopkgtest suite does that)

Dridi replied:
My stance on it was clearly nope with the first version of the script (see https://github.com/varnishcache/pkg-varnish-cache/issues/76#issuecomment-306676097 for example) but as I was about to reply the same to you, it occurred to me that reloading several VCLs in the same second now makes sense if you want to reload several labels at once.

Thanks!

Dridi replied:
Oh great, `%N` is not portable, looking for an alternative.

Dridi replied:
For reference, I first checked [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/date.html) and upon not seeing `%N` looked at [FreeBSD](http://www.unix.com/man-page/FreeBSD/1/date/) to see whether I could close my eyes on this detail, but I can't...

ssm commented:
Dridi Boukelmoune <notifications@github.com> writes:

> @ssm good to know! And while I have your attention, can you test the
> new version of the script on Jessie? And if I'm not stretching too
> hard, Ubuntu 14.04 too?

I've not got either of those handy as test beds, sorry.

-- 
Stig Sandbeck Mathisen

Dridi commented:
> I've not got either of those handy as test beds, sorry.

It's fine, I'll check Debian 8 as tested now, making progress :)

ssm commented:
Dridi Boukelmoune <notifications@github.com> writes:

> Oh great, `%N` is not portable, looking for an alternative.

Add the current pid instead?

-- 
Stig Sandbeck Mathisen
https://fnord.no/

ssm commented:
for instance:
```
vcl_reload_name() {
  date +"reload_%Y%m%d_%H%M%S_$$"
}
```

Dridi commented:
I guess it's time to share my love of POSIX, it's been a while since I last complained... I couldn't find anything readily available in shell to get the current time with milliseconds resolution. I looked at the set of standard commands, the `sh` built-ins, `awk` functions, and I ran out of ideas.

@ssm, so far your PID suggestion is the best option.

ssm commented:
I looked at the same things, and stopped when I got to the point where reading /dev/urandom with "dd" started to look good, and then use shell to transform it into a few random bytes Varnish would like.  Possibly fun, but not particularly elegant. :)

Dridi commented:
I finally ironed out all bugs and regressions I could find with the introduction of labels support, and I'm ready to start a test suite, unless...

> 19:26 < dridi> phk: around?
19:26 < phk> dridi, dinner in a few minutes, but until then
19:26 < dridi> this is regarding inclusion of a varnishreload script in varnish-cache
19:27 < dridi> I've been doing a lot of shell scripting lately and decided to strike this iron while it's hot
19:27 < dridi> so now I have a script that supports reloading labels
19:27 < dridi> I have personally tested it on linux with bash and dash, and people stepped up to test it on other linuxes
19:28 < dridi> and I tested it on freebsd and sunos
19:28 < phk> dridi, counter proposal:  Give VCL names version extensions and always use the higest version
19:28 < phk> dridi, "vcl.load foobar /etc/vcl.vcl"  -> foobar:0
19:28 < phk> dridi, "vcl.load foobar /etc/vcl.vcl"  -> foobar:1
19:28 < phk> dridi, "vcl.load foobar /etc/vcl.vcl"  -> foobar:2
19:28 < phk> dridi, vcl.use foobar:1
19:29 < phk> dridi, "vcl.use foobar"  -> use foobar:2
19:29 < dridi> vcl.label prod foobar
19:29 < dridi> vcl.load foobar ...
19:29 < phk> dinner
19:29 < dridi> is prod updated?
19:29 < dridi> ok
19:29 < dridi> I'm interested in the discard behavior

I guess this will be downstream-only, I'm not into lepidopterology.
